java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -196,8 +194,27 @@ public void visitContinueStatement(ContinueStatementTree tree) {
   @Override
   public void visitMethodInvocation(MethodInvocationTree tree) {
     MethodInvocationTreeImpl mit = (MethodInvocationTreeImpl) tree;
-    if(mit.isTypeSet()) {
-      // nothing special to deduce once we have set type of expression
+    if(mit.isTypeSet() && mit.symbol().isMethodSymbol()) {
+      JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) mit.symbol();
+      JavaType parametrizedTypeType = (JavaType) mit.symbolType();
+      if(parametrizedTypeType.isTagged(JavaType.PARAMETERIZED)) {
+        TypeSubstitution typeSubstitution = ((ParametrizedTypeJavaType) parametrizedTypeType).typeSubstitution.replaceKeys(methodSymbol.typeVariableTypes);
+        parametrizedTypeType = parametrizedTypeCache.getParametrizedTypeType((JavaSymbol.TypeJavaSymbol) parametrizedTypeType.symbol(), typeSubstitution);
```

Comment:
Rather than creating a new `parametrizedTypeType`, which uses type variables potentially totally unrelated (not defined as being its own type variables), I think it would be more interesting to only use this new `typeSubstitution` directly, and call `TypeSubstitutionSolver.applySubstitutionToFormalParameters(List<JavaType>, TypeSubstitution)` do infer formals.